crates/ironrdp-connector/src/credssp_sequence.rs

```diff
@@ -0,0 +1,214 @@
+use ironrdp_pdu::write_buf::WriteBuf;
+use ironrdp_pdu::{nego, PduHint};
+use sspi::credssp::{self, ClientState, CredSspClient};
+use sspi::generator::{Generator, NetworkRequest};
+use sspi::negotiate::ProtocolConfig;
+use sspi::KerberosConfig;
+
+use crate::{
+    ClientConnector, ClientConnectorState, ConnectorError, ConnectorErrorKind, ConnectorResult, ServerName, Written,
+};
+
+#[derive(Clone, Copy, Debug)]
+struct CredsspTsRequestHint;
+
+const CREDSSP_TS_REQUEST_HINT: CredsspTsRequestHint = CredsspTsRequestHint;
+
+impl PduHint for CredsspTsRequestHint {
+    fn find_size(&self, bytes: &[u8]) -> ironrdp_pdu::PduResult<Option<usize>> {
+        match sspi::credssp::TsRequest::read_length(bytes) {
+            Ok(length) => Ok(Some(length)),
+            Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => Ok(None),
+            Err(e) => Err(ironrdp_pdu::custom_err!("CredsspTsRequestHint", e)),
+        }
+    }
+}
+#[derive(Clone, Copy, Debug)]
+struct CredsspEarlyUserAuthResultHint;
+
+const CREDSSP_EARLY_USER_AUTH_RESULT_HINT: CredsspEarlyUserAuthResultHint = CredsspEarlyUserAuthResultHint;
+
+impl PduHint for CredsspEarlyUserAuthResultHint {
+    fn find_size(&self, _: &[u8]) -> ironrdp_pdu::PduResult<Option<usize>> {
+        Ok(Some(sspi::credssp::EARLY_USER_AUTH_RESULT_PDU_SIZE))
+    }
+}
+
+pub type CredSspProcessGenerator<'a> = Generator<'a, NetworkRequest, sspi::Result<Vec<u8>>, sspi::Result<ClientState>>;
+
+#[derive(Debug)]
+pub struct CredSspSequence {
+    client: CredSspClient,
+    next_request: Option<credssp::TsRequest>,
+    state: CredSSPState,
+    selected_protocol: nego::SecurityProtocol,
+}
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum CredSSPState {
```

Comment:
[Rust naming convention](https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case) is
```suggestion
pub(crate) enum CredsspState {
```
> In UpperCamelCase, acronyms and contractions of compound words count as one word: use Uuid rather than UUID, Usize rather than USize or Stdin rather than StdIn. In snake_case, acronyms and contractions are lower-cased: is_xid_start.